torch/nn/modules/module.py

```diff
@@ -551,6 +551,81 @@ def __repr__(self):
         tmpstr = tmpstr + ')'
         return tmpstr
 
+    def summary(self, input_size):
+        def register_hook(module):
+            def hook(module, input, output):
+                class_name = str(module.__class__).split('.')[-1].split("'")[0]
+                module_idx = len(summary)
+                m_key = '%s-%i' % (class_name, module_idx + 1)
+                summary[m_key] = OrderedDict()
+                summary[m_key]['input_shape'] = list(input[0].size())
+                summary[m_key]['input_shape'][0] = None
+                summary[m_key]['output_shape'] = list(output.size())
+                summary[m_key]['output_shape'][0] = None
+
+                params = 0
+                # iterate through parameters and count num params
+                for name, p in module._parameters.items():
+                    params += torch.numel(p.data)
+                    if p.requires_grad:
+                        summary[m_key]['trainable'] = True
+                    else:
+                        summary[m_key]['trainable'] = False
+
+                summary[m_key]['nb_params'] = params
+
+            if not isinstance(module, torch.nn.Sequential) and \
+               not isinstance(module, torch.nn.ModuleList) and \
+               not (module == self):
+                hooks.append(module.register_forward_hook(hook))
+
+        # check if there are multiple inputs to the network
+        if isinstance(input_size[0], (list, tuple)):
+            x = [Variable(torch.rand(1, *in_size)) for in_size in input_size]
+        else:
+            x = Variable(torch.randn(1, *input_size))
+
+        # create properties
+        summary = OrderedDict()
+        hooks = []
+        # register hook
+        self.apply(register_hook)
+        # make a forward pass
+        self(x)
+        # remove these hooks
+        for h in hooks:
+            h.remove()
+
+        # print out neatly
+        names = list(self._modules.keys())
+        col_width = 25  # should be >= 12
+        summary_width = 61
+
+        def crop(s):
+            return s[:col_width] if len(s) > col_width else s
+
+        print('_' * summary_width)
+        print('{0: <{3}} {1: <{3}} {2: <{3}}'.format(
+            'Layer (type)', 'Output Shape', 'Param #', col_width))
+        print('=' * summary_width)
+        total_params = 0
+        trainable_params = 0
+        for (i, l_type), l_name in zip(enumerate(summary), names):
```

Comment:
I see. I fixed the issue - I recursively go down to the layers now. To avoid a confusing output, I do not display a separate layer row for container modules, but simply prefix the container name (in this case `mod`) before the actual layers themselves. The summary of the model you posted above is displayed as:

![Summary](https://i.imgur.com/nKva1Kq.png)